numpy/core/src/multiarray/strfuncs.c

```diff
@@ -197,3 +197,30 @@ array_str(PyArrayObject *self)
     }
     return s;
 }
+
+NPY_NO_EXPORT PyObject *
+array_format(PyArrayObject *self, PyObject *args)
+{
+    PyObject *format;
+    if (!PyArg_ParseTuple(args, "O:__format__", &format))
+        return NULL;
+
+    /* 0d arrays - forward to the scalar type */
+    if (PyArray_NDIM(self) == 0) {
+        PyObject *item = PyArray_ToScalar(PyArray_DATA(self), self);
+        PyObject *res;
+        if (item == NULL) {
+            return NULL;
+        }
+        res = PyObject_Format(item, format);
+        Py_DECREF(item);
+        return res;
+    }
+    /* Everything else - use the builtin */
```

Comment:
What does the builtin do here?

Reply:
It keeps the arbitrary and python-dependent behavior that we had before this patch.

Most importantly it handles `'{}'.format` in the typical way. On python 3, it throws an exception if a format string is given, while on python 2 it evaluates `'{:theformat}'.format(str(arr))`.